pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go

```diff
@@ -1841,242 +1841,310 @@ func getNumberOfPodInterfaces(spec *v1.VirtualMachineInstanceSpec) int {
-				causes = append(causes, metav1.StatusCause{
-					Type:    metav1.CauseTypeFieldValueNotSupported,
-					Message: fmt.Sprintf("IOThreads are not supported for disks on a %s bus", bus),
-					Field:   field.Child("domain", "devices", "disks").Index(idx).String(),
-				})
-			}
-		}
+func getDiskType(disk v1.Disk) string {
+	switch {
+	case disk.Disk != nil:
+		return "disk"
+	case disk.LUN != nil:
+		return "lun"
+	case disk.CDRom != nil:
+		return "cdrom"
+	default:
+		return ""
+	}
+}
 
-		// Verify serial number is made up of valid characters for libvirt, if provided
-		if disk.Serial != "" && !isValidExpression(disk.Serial) {
+func validateBusSupport(field *k8sfield.Path, idx int, disk v1.Disk) []metav1.StatusCause {
+	var causes []metav1.StatusCause
+	bus := getDiskBus(disk)
+	diskType := getDiskType(disk)
+	if bus == "" {
+		return causes
+	}
+	switch bus {
+	case "ide":
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: "IDE bus is not supported",
+			Field:   field.Index(idx).Child(diskType, "bus").String(),
+		})
+	case v1.DiskBusVirtio:
+		// special case. virtio is incompatible with CD-ROM for q35 machine types
+		if diskType == "cdrom" {
 			causes = append(causes, metav1.StatusCause{
 				Type:    metav1.CauseTypeFieldValueInvalid,
-				Message: fmt.Sprintf("%s must be made up of the following characters [A-Za-z0-9_.+-], if specified", field.Index(idx).String()),
-				Field:   field.Index(idx).Child("serial").String(),
+				Message: fmt.Sprintf("Bus type %s is invalid for CD-ROM device", bus),
+				Field:   field.Index(idx).Child("cdrom", "bus").String(),
 			})
 		}
-
-		// Verify serial number is within valid length, if provided
-		if disk.Serial != "" && len([]rune(disk.Serial)) > maxStrLen {
+	case v1.DiskBusSATA:
+		// sata disks (in contrast to sata cdroms) don't support readOnly
+		if disk.Disk != nil && disk.Disk.ReadOnly {
 			causes = append(causes, metav1.StatusCause{
 				Type:    metav1.CauseTypeFieldValueInvalid,
-				Message: fmt.Sprintf("%s must be less than or equal to %d in length, if specified", field.Index(idx).String(), maxStrLen),
-				Field:   field.Index(idx).Child("serial").String(),
+				Message: fmt.Sprintf("%s hard-disks do not support read-only.", bus),
+				Field:   field.Index(idx).Child("disk", "bus").String(),
 			})
 		}
+	case v1.DiskBusSCSI, v1.DiskBusUSB:
+		break
+	default:
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("%s is set with an unrecognized bus %s, must be one of: %v", field.Index(idx).String(), bus, supportedBuses),
+			Field:   field.Index(idx).Child(diskType, "bus").String(),
+		})
+	}
+	// Reject defining DedicatedIOThread to a disk without VirtIO bus since this configuration
+	// is not supported in libvirt.
+	isIOThreadsWithUnsupportedBus := disk.DedicatedIOThread != nil && *disk.DedicatedIOThread && bus != v1.DiskBusVirtio
```

Comment:
nit: you can move this under the DiskBusVirtio case

Reply:
I think this can't be done as the condition is checking for `bus != v1.DiskBusVirtio`. I could add this block in all cases other than v1.DiskBusVirtio but I felt that might lead to code duplication, so kept it here.

Reply:
you are right I misread it

Reply:
Might be okay to inline this into L1985